third_party/move/move-vm/types/src/loaded_data/ty_args_fingerprint.rs

```diff
@@ -0,0 +1,319 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::loaded_data::runtime_types::Type;
+use once_cell::sync::Lazy;
+use std::hash::{BuildHasher, Hash, Hasher};
+
+/// Fingerprint over a fully-resolved type-argument vector.
```

Comment:
What properties does it have? Is it a perfect hash? What's the complexity to compute it? I

Suggest to have some `//!` level conceptual doc at the file beginning.

From looking at the code, it seems to be definitely not a perfect hash, so I wonder how this works as an index for frame caches and why this would be correct. 

This argument is specifically important for security review since we know how dangerous wrong frame caches can be. Perhaps it would be good to first write down (AIP style) what this is doing?

Reply:
Indeed - this is imperfect hash so we should have collision resolution. I was thinking this or interning table and I think interning is probably much better and is more aligned with future type work + is collision-free. Updated the PR but need to recheck performance first and add documentation as you pointed out👍